clients/roscpp/src/libros/spinner.cpp

```diff
@@ -30,10 +30,90 @@
 #include "ros/callback_queue.h"
 
 #include <boost/thread/thread.hpp>
-#include <boost/thread/recursive_mutex.hpp>
+#include <boost/thread/mutex.hpp>
 
 namespace {
-  boost::recursive_mutex spinmutex;
+
+/** class to monitor running single-threaded spinners.
+ *
+ *  Calling the callbacks of a callback queue _in order_, requires a unique SingleThreadedSpinner
+ *  spinning on the queue. Other threads accessing the callback queue will probably intercept execution order.
+
+ *  To avoid multiple SingleThreadedSpinners (started from different threads) to operate on the same callback queue,
+ *  this class stores a map of all spinned callback queues.
+ *  If the spinner is single threaded, the corresponding thread-id is stored in the map
+ *  and if other threads will try to spin the same queue, an error message is issued.
+ *
+ *  If the spinner is multi-threaded, the stored thread-id is NULL and future SingleThreadedSpinners
+ *  should not spin this queue.
+ */
+struct SpinnerMonitor
+{
+  /* store spinner information per callback queue:
+     Only as unique single-threaded spinner is allowed and we store its thread id.
+     As multiple multi-threaded spinners are allowed in parallel, their number is counted.
+  */
+  struct Entry
+  {
+    Entry(const boost::thread::id &tid) : tid(tid) {}
+
+    boost::thread::id tid; // thread id of single-threaded spinner
+    unsigned int num_multi_threaded; // number of multi-threaded spinners
```

Comment:
Indeed once things are in separate threads you're at the whim of the scheduler. I was more making the point that there's a level that we can make sure that the multiple servicing callback queues coordinate cleanly and do things like immediately invoke the callbacks they pop off the queue instead of popping a batch of the incoming queue and then internally queuing them for invocation such that the other multithreaded callback queue gets a batch of later messages and starts invoking callback from the 2nd batch while the first multithreaded callback queue is processing threads from the first batch. I don't think that's the case in our implementation. So I'm not objecting to the current patch but want to make sure that we are doing our best effort to make sure invoke callbacks in order up to the limit of our control.